### ISSUE TITLE
[Question] Question title自己训练的GO2代码部署

### ISSUE THREAD

**Issue by HaodongHuang-W** [fan-ziqi/rl_sar#72]

**Comment by HaodongHuang-W** [comment]
哥，我在用你的代码的过程中，我在go2的model下面新建了代码

![Image](https://github.com/user-attachments/assets/87f17a2e-a149-4340-8fd2-940c14873843)

![Image](https://github.com/user-attachments/assets/11a23a78-34ee-4909-87b7-a287d42a9809)

![Image](https://github.com/user-attachments/assets/d0a6ba56-b50b-40b8-890e-81844fd02a5b)
sim2sim的时候一直报错，然后其他的himloco都可以，就我的这个不可以，报错如下

![Image](https://github.com/user-attachments/assets/0f13d8a8-9062-4447-a827-9bbd128cbd66)
`go2/BYOL:
  model_name: "byol.pt"
  framework: "isaacgym"
  dt: 0.005
  decimation: 4
  num_observations: 45
#   observations: ["commands", "ang_vel", "gravity_vec", "dof_pos", "dof_vel", "actions"]
  observations: ["ang_vel", "gravity_vec", "commands", "dof_pos", "dof_vel", "actions"]
  observations_history: [0, 1, 2, 3, 4, 5,6,7,8,9]  # 0 is the latest observation
  clip_obs: 100.0
  clip_actions_lower: [-100, -100, -100,
                       -100, -100, -100,
                       -100, -100, -100,
                       -100, -100, -100]
  clip_actions_upper: [100, 100, 100,
                       100, 100, 100,
                       100, 100, 100,
                       100, 100, 100]
  rl_kp: [40, 40, 40,
          40, 40, 40,
          40, 40, 40,
          40, 40, 40]
  rl_kd: [1, 1, 1,
          1, 1, 1,
          1, 1, 1,
          1, 1, 1]
  fixed_kp: [60, 60, 60,
             60, 60, 60,
             60, 60, 60,
             60, 60, 60]
  fixed_kd: [5, 5, 5,
             5, 5, 5,
             5, 5, 5,
             5, 5, 5]
  num_of_dofs: 12
  action_scale: [0.125, 0.25, 0.25,
                 0.125, 0.25, 0.25,
                 0.125, 0.25, 0.25,
                 0.125, 0.25, 0.25]
  wheel_indices: []
  lin_vel_scale: 2.0
  ang_vel_scale: 0.25
  dof_pos_scale: 1.0
  dof_vel_scale: 0.05
  commands_scale: [2.0, 2.0, 0.25]
#   torque_limits: [33.5, 33.5, 33.5,
#                   33.5, 33.5, 33.5,
#                   33.5, 33.5, 33.5,
#                   33.5, 33.5, 33.5]

  torque_limits: [23.5, 23.5, 23.5,
                  23.5, 23.5, 23.5,
                  23.5, 23.5, 23.5,
                  23.5, 23.5, 23.5]

  default_dof_pos: [ 0.1000, 0.8000, -1.5000,
                    -0.1000, 0.8000, -1.5000,
                     0.1000, 1.0000, -1.5000,
                    -0.1000, 1.0000, -1.5000]
  # joint_controller_names: ["FL_hip_controller", "FL_thigh_controller", "FL_calf_controller",
  #                          "FR_hip_controller", "FR_thigh_controller", "FR_calf_controller",
  #                          "RL_hip_controller", "RL_thigh_controller", "RL_calf_controller",
  #                          "RR_hip_controller", "RR_thigh_controller", "RR_calf_controller"]

  joint_controller_names: ["FR_hip_controller", "FR_thigh_controller", "FR_calf_controller",
                           "FL_hip_controller", "FL_thigh_controller", "FL_calf_controller",
                           "RR_hip_controller", "RR_thigh_controller", "RR_calf_controller",
                           "RL_hip_controller", "RL_thigh_controller", "RL_calf_controller"]
#   command_mapping: [3, 4, 5, 0, 1, 2, 9, 10, 11, 6, 7, 8]
#   state_mapping: [3, 4, 5, 0, 1, 2, 9, 10, 11, 6, 7, 8]
  command_mapping: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
  state_mapping: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]`

**Comment by HaodongHuang-W** [comment]
然后问了deepseek肯torch不兼容

**Comment by fan-ziqi** [comment]
拉一下最新代码，更新了错误提示

**Comment by HaodongHuang-W** [comment]
好的哥

**Comment by HaodongHuang-W** [comment]
哥，上面的问题我解决了，感谢，目前有一个新的问题
[INFO] RL Controller x:0.00 y:0.00 yaw:0.00terminate called after throwing an instance of 'c10::Error'
  what():  forward() is missing value for argument 'obs_hist'. Declaration: forward(__torch__.modules.actor_critic.MlpBYOLActor self, Tensor obs, Tensor obs_hist) -> Tensor
Exception raised from checkAndNormalizeInputs at ../aten/src/ATen/core/function_schema_inl.h:417 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7f9cc25a005b in /home/hhd/libtorch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xbf (0x7f9cc259af6f in /home/hhd/libtorch/lib/libc10.so)
frame #2: void c10::FunctionSchema::checkAndNormalizeInputs<c10::Type>(std::vector<c10::IValue, std::allocator<c10::IValue> >&, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, c10::IValue> > > const&) const + 0x7d1 (0x7f9cc3baa161 in /home/hhd/libtorch/lib/libtorch_cpu.so)
frame #3: torch::jit::Method::operator()(std::vector<c10::IValue, std::allocator<c10::IValue> >, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, c10::IValue> > > const&) const + 0x173 (0x7f9cc6badb53 in /home/hhd/libtorch/lib/libtorch_cpu.so)
frame #4: torch::jit::Module::forward(std::vector<c10::IValue, std::allocator<c10::IValue> >, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, c10::IValue> > > const&) + 0xc3 (0x55a4e374f6cb in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #5: RL_Sim::Forward() + 0x24a (0x55a4e3742b4e in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #6: RL_Sim::RunModel() + 0x771 (0x55a4e3742509 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #7: void std::__invoke_impl<void, void (RL_Sim::*&)(), RL_Sim*&>(std::__invoke_memfun_deref, void (RL_Sim::*&)(), RL_Sim*&) + 0x6a (0x55a4e378afaa in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #8: std::__invoke_result<void (RL_Sim::*&)(), RL_Sim*&>::type std::__invoke<void (RL_Sim::*&)(), RL_Sim*&>(void (RL_Sim::*&)(), RL_Sim*&) + 0x4a (0x55a4e378ac64 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #9: void std::_Bind<void (RL_Sim::*(RL_Sim*))()>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) + 0x5b (0x55a4e378a9b1 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #10: void std::_Bind<void (RL_Sim::*(RL_Sim*))()>::operator()<, void>() + 0x37 (0x55a4e378a425 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #11: std::_Function_handler<void (), std::_Bind<void (RL_Sim::*(RL_Sim*))()> >::_M_invoke(std::_Any_data const&) + 0x24 (0x55a4e3789d4b in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #12: std::function<void ()>::operator()() const + 0x36 (0x55a4e375ac00 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #13: LoopFunc::loop() + 0x51 (0x55a4e3750a65 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #14: void std::__invoke_impl<void, void (LoopFunc::*)(), LoopFunc*>(std::__invoke_memfun_deref, void (LoopFunc::*&&)(), LoopFunc*&&) + 0x6a (0x55a4e3791d12 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #15: std::__invoke_result<void (LoopFunc::*)(), LoopFunc*>::type std::__invoke<void (LoopFunc::*)(), LoopFunc*>(void (LoopFunc::*&&)(), LoopFunc*&&) + 0x4a (0x55a4e3791681 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #16: void std::thread::_Invoker<std::tuple<void (LoopFunc::*)(), LoopFunc*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) + 0x47 (0x55a4e3790683 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #17: std::thread::_Invoker<std::tuple<void (LoopFunc::*)(), LoopFunc*> >::operator()() + 0x2b (0x55a4e378f7f5 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #18: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (LoopFunc::*)(), LoopFunc*> > >::_M_run() + 0x20 (0x55a4e378dc18 in /home/hhd/rl_sar/devel/.private/rl_sar/lib/rl_sar/rl_sim)
frame #19: <unknown function> + 0xd6df4 (0x7f9cc2436df4 in /lib/x86_64-linux-gnu/libstdc++.so.6)
frame #20: <unknown function> + 0x8609 (0x7f9cd9e78609 in /lib/x86_64-linux-gnu/libpthread.so.0)
frame #21: clone + 0x43 (0x7f9cc2272353 in /lib/x86_64-linux-gnu/libc.so.6)
就是模型的输入对不上，
哥在我的训练的actor的网络中
`    def forward(self, obs, obs_hist, return_projs=False):

        # print("obs:",obs.shape)
        # print("obs_hist_full:",obs_hist.shape)
        obs, obs_hist = self.normalize(obs, obs_hist)
        obs_hist_full = torch.cat([obs_hist[:,1:,:], obs.unsqueeze(1)], dim=1)

`中我需要把obs和obs_hist都输入给policy，但是
`torch::Tensor RL_Sim::Forward()
{
    torch::autograd::GradMode::set_enabled(false);

    torch::Tensor clamped_obs = this->ComputeObservation();

    torch::Tensor actions;
    if (this->params.observations_history.size() != 0)
    {
        
        this->history_obs = this->history_obs_buf.get_obs_vec(this->params.observations_history);
        actions = this->model.forward({clamped_obs,this->history_obs}).toTensor();
        this->history_obs_buf.insert(clamped_obs);
    }
    else
    {
        actions = this->model.forward({clamped_obs}).toTensor();
    }

    if (this->params.clip_actions_upper.numel() != 0 && this->params.clip_actions_lower.numel() != 0)
    {
        return torch::clamp(actions, this->params.clip_actions_lower, this->params.clip_actions_upper);
    }
    else
    {
        return actions;
    }
}
}`我的forward，我加入了还是报错，说没有历史观测量，对了哥，我把observations_history: [9,8,7,6,5,4,3,2,1,0] ，这个9写在前面是不是代表我最后的观测量是最新的。感谢哥

**Comment by fan-ziqi** [comment]
observations_history的顺序在yaml中有说明
